pageserver/src/tenant_mgr.rs

```diff
@@ -596,6 +596,10 @@ fn collect_timelines_for_tenant(
     let timelines_dir = config.timelines_path(&tenant_id);
 
     let mut tenant_timelines = HashMap::new();
+    if !timelines_dir.as_path().is_dir() {
+        return Ok((tenant_id, tenant_timelines));
```

Comment:
Ok so just to understand it:
-  `local_tenant_timeline_files`: We load the tenant with the timelines, if  the `tenant_timelines_dir` does not exists then we return the tenant with an empty collection of timelines.
- `attach_local_tenants`: We call `write_tenants` for each tenant that it's not already present, if the `tenant_timelines_dir` does not exists then its 'Broken' (also can be `Broken` because of an error).


We validate the dir existence two times.

Do you want to load and attach all non `Broken` tenants and then `write_tenant` for every `Broken` one?
Whats the other part needed? In #2552 i see `Timelines would require the same treatment` is that the other part? What would be the behaviour in that case?

Thanks for the review!

Reply:
Yes, you understand it right, there are two those steps exactly.
I think, the 2nd check is not needed and what we need to do is to determine as many issues as we can from the local FS structure: then, we pass on (from the first bullet, presumably) a list of all FS-processed entities, with some of them already known to be broken (such as tenants with no `timelines/` dir).

We add such broken tenants into memory as broken right away and never bother to check local FS structure for other tenants, considering them to be ok and loading them further.

Reply:
Ok so let me know what do you think with the new changes, i need to polish them but first i wanted to know ur opinion.

- `local_tenant_timeline_files ` now returns `TenantStatus`: We need a separate collection of broken and non-broken tenants so i can return a tuple or create a struct only for readability. If we do not need to separate tenants for early writing of broken ones we could return to the previous flow of returning a collection of processed tenants.
- `load_local_tenant`removal: Totally opinionated i can change it.

Thanks!

Reply:
That's definitely a step forward, I like the direction.
I have some doubts that it's worth creating two separate structs for the task, I think we should abandon `TenantTimelineValues` here and redefine the types to smth. like

```rust
fn foo(tenants_to_attach: HashMap<TenantId, TenantAttachData>) -> .....

enum TenantAttachData {
    Ready(HashMap<TimelineId, TimelineMetadata>),
    Broken(anyhow::Error),
}
```

Presumably, no extra structures is needed them and we can match over the entries once in the end, before deciding to attach or not.
It might turn out that your current approach is more ergonomical (I kind of like it how batched those entities are now), but then let's consider different names for both new entities, IMO neither of the current ones is descriptive.

Reply:
Yup that's clean... Awesome, thanks!
I added to that `Ready` the `HashSet<PathBuf>` because of remote storage sync and also i wanted to be explicit about broken-ready in [there](https://github.com/neondatabase/neon/pull/2575/commits/476e290536054df288d2bf47ad3963d284075273#diff-0d4539a7193c24a7f3349df105b8bf31a2767e74bcd4898532cb7d44838bf499R75) just to not mix timelines and tenant status inside `fn spawn_storage_sync_task `. 

Let me know what you think please, i appreciate your feedback!